src/sim.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+import math
+import numpy as np
+import pandas as pd
+import salabim as sim
+import sys
+
+
+#  based on SimPy example model
+class Bus(sim.Component):
+    """
+    TODO
+    """
+    def __init__(self, start_time, end_time, trip_power, charging_rate, *args, **kwargs):
+        sim.Component.__init__(self, *args, **kwargs)
+        self.start_time    = start_time
+        self.end_time      = end_time
+        self.trip_power    = trip_power
+        self.charging_rate = charging_rate
+
+    def process(self):
+        """TODO
+        """
+        #Trip starts when this component is activated
+        yield self.request(bus_depot)
+        yield self.hold(till=self.end_time)
+        yield self.request((energy, self.trip_power))
+        yield self.hold(duration=self.trip_power/self.charging_rate*3600)
+        self.release(bus_depot)
+
+
+
+class Dispatcher:
+    def __init__(self, input_file, battery_cap_kwh=100, kwh_per_km=2, charging_rate=150):
+        self.gtfs            = pd.read_pickle(input_file, compression='infer')
+        self.gtfs            = self.gtfs.sort_values(['block_id', 'start_arrival_time'])
+        self.gtfs            = self.gtfs[~self.gtfs.trip_id.str.contains('RAIL')]
+        self.buses           = []
+        self.battery_cap_kwh = battery_cap_kwh
+        self.kwh_per_km      = kwh_per_km
+        self.charging_rate   = charging_rate
+        self.trip_seg_count  = 0
+        for block_id, group in self.gtfs.groupby(['block_id']):
+            self.schedule_block(block_id, group)
+
+    def schedule_block(self, block_id, trips):
+        trips = trips.copy()
+        distance_to_trip    = 5 #km. TODO: make this the maximum of the distance to the start and end
+        travel_time_to_trip = 0.2*3600 #seconds
+        maxdist             = self.battery_cap_kwh/self.kwh_per_km-2*distance_to_trip
+        trips['cumdist']  = trips.distance.cumsum()
+        trips['trip_seg'] = (trips.cumdist/1000/maxdist).astype(np.int)
+        # print(trips[['trip_headsign','start_arrival_time','distance','cumdist','trip_seg']])
+        for trip_seg, seg_group in trips.groupby('trip_seg'):
+            trip_power      = (seg_group.iloc[-1].cumdist-seg_group.iloc[0].cumdist+2*distance_to_trip)*self.kwh_per_km
+            trip_start_time = seg_group.iloc[0].start_arrival_time-travel_time_to_trip
+            trip_end_time   = seg_group.iloc[-1].end_arrival_time+travel_time_to_trip
+            new_bus         = Bus(trip_start_time,trip_end_time, trip_power, self.charging_rate, at=trip_start_time)
+            self.trip_seg_count += 1
+
```

Comment:
I need help understanding what is going on here. I dont quite understand when and how new buses are being made if a bus cant serve the full number of trips it was originally supposed to serve. 

I also dont understand how any charging is happening. Are buses going back to the depot and being reassigned, or are we simply running buses dry and then making a new bus?

Reply:
See above:
```
trips['trip_seg'] = (trips.cumdist/1000/maxdist).astype(np.int)
```
breaks the trip into portions which can be served by single buses. These buses are then created in this chunk. Buses return to the depot to get charged (see the `Bus` class).